app/questionnaire/router.py

```diff
@@ -1,67 +1,141 @@
+
+        if self._schema.is_block_list_collector_child(block_id):
+            # If this is a list collector sub block, return the collector in the previous link
+            block = self._schema.get_block(block_id)
+            return Location(block_id=block['parent_id'])
+
+        location_index = routing_path.index(location)
+
+        if location_index != 0:
+            return routing_path[location_index - 1]
 
-        return first_incomplete_location or self._routing_path[-1]
+        return None
+
+    def get_first_incomplete_location_in_survey(self):
+        incomplete_section_ids = self._get_incomplete_section_ids()
+
+        if incomplete_section_ids:
+            first_incomplete_section = self._schema.get_section(
+                incomplete_section_ids[0]
+            )
+            section_routing_path = path_finder.routing_path(first_incomplete_section)
+            location = path_finder.get_first_incomplete_location_on_path(
+                section_routing_path
+            )
+            if location:
+                return location
 
-    def _get_survey_redirect_location(self):
+        all_section_ids = self._schema.get_section_ids()
+        last_block_id = self._schema.get_last_block_id_for_section(all_section_ids[-1])
+        return Location(block_id=last_block_id)
+
+    def is_survey_complete(self):
+        incomplete_section_ids = self._get_incomplete_section_ids()
+
+        if incomplete_section_ids:
+            if len(incomplete_section_ids) > 1:
+                return False
+            if self._does_section_only_contain_summary(incomplete_section_ids[0]):
+                return True
+            return False
+
+        return True
+
+    def _get_allowable_path(self, routing_path):
         """
-        Ensure that the user can't access the end of the survey if the survey is not complete.
-        :return: Location to direct to if current location is not accessible.
+        The allowable path is the completed path plus the next location
         """
-        if not self._is_current_location_valid():
-            return None
+        allowable_path = []
+        for location in routing_path:
+            if location in self._completed_store.block_locations:
+                allowable_path.append(location)
+            else:
+                allowable_path.append(location)
+                return allowable_path
 
-        expected_location = self._get_first_incomplete_location_in_survey()
-        block = self._schema.get_block(self._current_location.block_id)
-        cannot_access_survey_end = self._current_location != expected_location
+        return allowable_path
 
-        if block['type'] in ['Confirmation', 'Summary'] and cannot_access_survey_end:
-            return expected_location
+    def _get_incomplete_section_ids(self):
+        all_section_ids = self._schema.get_section_ids()
+        completed_sections = {section for section in self._completed_store.sections}
```

Comment:
Can we get rid of this var and avoid the extra comprehension?
`self._completed_store.sections` are unique anyway but that doesn't matter anyway?

Reply:
Heh, good spot. I've removed the `completed_sections` variable.